### PR TITLE
Bugfix/commons 101 illegal cyclic reference

### LINKED ISSUE #101: scala.reflect.internal.Symbols$CyclicReference: illegal cyclic reference involving class InterfaceAudience
printing stack trace fails because the cause of the exception is cyclic. This is info from debbuger:

exception.getClass
```
class java.lang.AssertionError
```

exception.detailMessage
```
assertion failed: no symbol could be loaded from interface org.apache.hadoop.classification.InterfaceAudience$Public in class InterfaceAudience with name Public and classloader sun.misc.Launcher$AppClassLoader@18b4aac2
```

exception.stackTrace
```
0 = {StackTraceElement@23667} "scala.reflect.internal.SymbolTable.throwAssertionError(SymbolTable.scala:185)"
1 = {StackTraceElement@23668} "scala.reflect.runtime.JavaMirrors$JavaMirror.classToScala1(JavaMirrors.scala:1068)"
2 = {StackTraceElement@23669} "scala.reflect.runtime.JavaMirrors$JavaMirror.$anonfun$classToScala$1(JavaMirrors.scala:1026)"
3 = {StackTraceElement@23670} "scala.reflect.runtime.JavaMirrors$JavaMirror.$anonfun$toScala$1(JavaMirrors.scala:137)"
4 = {StackTraceElement@23671} "scala.reflect.runtime.TwoWayCaches$TwoWayCache.$anonfun$toScala$1(TwoWayCaches.scala:50)"
5 = {StackTraceElement@23672} "scala.reflect.runtime.TwoWayCaches$TwoWayCache.toScala(TwoWayCaches.scala:46)"
6 = {StackTraceElement@23673} "scala.reflect.runtime.JavaMirrors$JavaMirror.toScala(JavaMirrors.scala:135)"
7 = {StackTraceElement@23674} "scala.reflect.runtime.JavaMirrors$JavaMirror.classToScala(JavaMirrors.scala:1026)"
8 = {StackTraceElement@23675} "scala.reflect.runtime.JavaMirrors$JavaMirror$JavaAnnotationProxy.<init>(JavaMirrors.scala:203)"
9 = {StackTraceElement@23676} "scala.reflect.runtime.JavaMirrors$JavaMirror$JavaAnnotationProxy$.apply(JavaMirrors.scala:202)"
10 = {StackTraceElement@23677} "scala.reflect.runtime.JavaMirrors$JavaMirror$JavaAnnotationProxy$.apply(JavaMirrors.scala:202)"
11 = {StackTraceElement@23678} "scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)"
12 = {StackTraceElement@23679} "scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)"
13 = {StackTraceElement@23680} "scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)"
14 = {StackTraceElement@23681} "scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)"
15 = {StackTraceElement@23682} "scala.collection.TraversableLike.map(TraversableLike.scala:286)"
16 = {StackTraceElement@23683} "scala.collection.TraversableLike.map$(TraversableLike.scala:279)"
17 = {StackTraceElement@23684} "scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:198)"
18 = {StackTraceElement@23685} "scala.reflect.runtime.JavaMirrors$JavaMirror.scala$reflect$runtime$JavaMirrors$JavaMirror$$copyAnnotations(JavaMirrors.scala:724)"
19 = {StackTraceElement@23686} "scala.reflect.runtime.JavaMirrors$JavaMirror$FromJavaClassCompleter.load(JavaMirrors.scala:779)"
20 = {StackTraceElement@23687} "scala.reflect.runtime.JavaMirrors$JavaMirror$FromJavaClassCompleter.complete(JavaMirrors.scala:790)"
21 = {StackTraceElement@23688} "scala.reflect.internal.Symbols$Symbol.completeInfo(Symbols.scala:1551)"
22 = {StackTraceElement@23689} "scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1514)"
23 = {StackTraceElement@23690} "scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anon$7.scala$reflect$runtime$SynchronizedSymbols$SynchronizedSymbol$$super$info(SynchronizedSymbols.scala:203)"
24 = {StackTraceElement@23691} "scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol.$anonfun$info$1(SynchronizedSymbols.scala:158)"
25 = {StackTraceElement@23692} "scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol.info(SynchronizedSymbols.scala:149)"
26 = {StackTraceElement@23693} "scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol.info$(SynchronizedSymbols.scala:158)"
27 = {StackTraceElement@23694} "scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anon$7.info(SynchronizedSymbols.scala:203)"
28 = {StackTraceElement@23695} "scala.reflect.internal.Symbols$Symbol.isDerivedValueClass(Symbols.scala:825)"
29 = {StackTraceElement@23696} "scala.reflect.runtime.JavaMirrors$JavaMirror$DerivedValueClassMetadata.<init>(JavaMirrors.scala:308)"
30 = {StackTraceElement@23697} "scala.reflect.runtime.JavaMirrors$JavaMirror$JavaVanillaMethodMirror0.<init>(JavaMirrors.scala:401)"
31 = {StackTraceElement@23698} "scala.reflect.runtime.JavaMirrors$JavaMirror.scala$reflect$runtime$JavaMirrors$JavaMirror$$mkMethodMirror(JavaMirrors.scala:363)"
32 = {StackTraceElement@23699} "scala.reflect.runtime.JavaMirrors$JavaMirror$JavaInstanceMirror.reflectMethod(JavaMirrors.scala:290)"
33 = {StackTraceElement@23700} "scala.reflect.runtime.JavaMirrors$JavaMirror$JavaInstanceMirror.reflectMethod(JavaMirrors.scala:275)"
34 = {StackTraceElement@23701} "za.co.absa.commons.reflect.ReflectionUtils$.$anonfun$extractFieldValue$5(ReflectionUtils.scala:143)"
35 = {StackTraceElement@23702} "scala.Option.map(Option.scala:230)"
36 = {StackTraceElement@23703} "za.co.absa.commons.reflect.ReflectionUtils$.reflectClass$1(ReflectionUtils.scala:141)"
37 = {StackTraceElement@23704} "za.co.absa.commons.reflect.ReflectionUtils$.reflectClassHierarchy$1(ReflectionUtils.scala:112)"
38 = {StackTraceElement@23705} "za.co.absa.commons.reflect.ReflectionUtils$.extractFieldValue(ReflectionUtils.scala:169)"
39 = {StackTraceElement@23706} "za.co.absa.commons.reflect.ReflectionUtils$.extractFieldValue(ReflectionUtils.scala:184)"
40 = {StackTraceElement@23707} "za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin$GeneratedEvaluatorClass$12$1.$anonfun$invoke$1(FileToCompile.scala:165)"
41 = {StackTraceElement@23708} "scala.util.Try$.apply(Try.scala:213)"
42 = {StackTraceElement@23709} "za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin$GeneratedEvaluatorClass$12$1.invoke(FileToCompile.scala:165)"
43 = {StackTraceElement@23710} "za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin.extractSourceIdFromDeltaTableV2(DataSourceV2Plugin.scala:162)"
44 = {StackTraceElement@23711} "za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin.za$co$absa$spline$harvester$plugin$embedded$DataSourceV2Plugin$$extractSourceIdFromTable(DataSourceV2Plugin.scala:152)"
45 = {StackTraceElement@23712} "za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin.za$co$absa$spline$harvester$plugin$embedded$DataSourceV2Plugin$$extractSourceIdFromRelation(DataSourceV2Plugin.scala:131)"
46 = {StackTraceElement@23713} "za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin$$anonfun$2.applyOrElse(DataSourceV2Plugin.scala:69)"
47 = {StackTraceElement@23714} "za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin$$anonfun$2.applyOrElse(DataSourceV2Plugin.scala:54)"
48 = {StackTraceElement@23715} "scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)"
49 = {StackTraceElement@23716} "scala.PartialFunction$Lifted.apply(PartialFunction.scala:228)"
50 = {StackTraceElement@23717} "scala.PartialFunction$Lifted.apply(PartialFunction.scala:224)"
51 = {StackTraceElement@23718} "za.co.absa.spline.harvester.builder.write.PluggableWriteCommandExtractor.asWriteCommand(PluggableWriteCommandExtractor.scala:43)"
52 = {StackTraceElement@23719} "za.co.absa.spline.harvester.LineageHarvester.$anonfun$tryExtractWriteCommand$1(LineageHarvester.scala:145)"
53 = {StackTraceElement@23720} "scala.util.Try$.apply(Try.scala:213)"
54 = {StackTraceElement@23721} "za.co.absa.spline.harvester.LineageHarvester.tryExtractWriteCommand(LineageHarvester.scala:145)"
55 = {StackTraceElement@23722} "za.co.absa.spline.harvester.LineageHarvester.harvest(LineageHarvester.scala:62)"
56 = {StackTraceElement@23723} "za.co.absa.spline.agent.SplineAgent$$anon$1.$anonfun$handle$1(SplineAgent.scala:85)"
57 = {StackTraceElement@23724} "za.co.absa.spline.agent.SplineAgent$$anon$1.withErrorHandling(SplineAgent.scala:94)"
58 = {StackTraceElement@23725} "za.co.absa.spline.agent.SplineAgent$$anon$1.handle(SplineAgent.scala:66)"
59 = {StackTraceElement@23726} "za.co.absa.spline.harvester.listener.QueryExecutionListenerDelegate.onSuccess(QueryExecutionListenerDelegate.scala:28)"
60 = {StackTraceElement@23727} "org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:158)"
61 = {StackTraceElement@23728} "org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:128)"
62 = {StackTraceElement@23729} "org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:117)"
63 = {StackTraceElement@23730} "org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:101)"
64 = {StackTraceElement@23731} "org.apache.spark.sql.util.ExecutionListenerBus.postToAll(QueryExecutionListener.scala:128)"
65 = {StackTraceElement@23732} "org.apache.spark.sql.util.ExecutionListenerBus.onOtherEvent(QueryExecutionListener.scala:140)"
66 = {StackTraceElement@23733} "org.apache.spark.scheduler.SparkListenerBus.doPostEvent(SparkListenerBus.scala:100)"
67 = {StackTraceElement@23734} "org.apache.spark.scheduler.SparkListenerBus.doPostEvent$(SparkListenerBus.scala:28)"
68 = {StackTraceElement@23735} "org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)"
69 = {StackTraceElement@23736} "org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)"
70 = {StackTraceElement@23737} "org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:117)"
71 = {StackTraceElement@23738} "org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:101)"
72 = {StackTraceElement@23739} "org.apache.spark.scheduler.AsyncEventQueue.super$postToAll(AsyncEventQueue.scala:105)"
73 = {StackTraceElement@23740} "org.apache.spark.scheduler.AsyncEventQueue.$anonfun$dispatch$1(AsyncEventQueue.scala:105)"
74 = {StackTraceElement@23741} "scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:23)"
75 = {StackTraceElement@23742} "scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)"
76 = {StackTraceElement@23743} "org.apache.spark.scheduler.AsyncEventQueue.org$apache$spark$scheduler$AsyncEventQueue$$dispatch(AsyncEventQueue.scala:100)"
77 = {StackTraceElement@23744} "org.apache.spark.scheduler.AsyncEventQueue$$anon$2.$anonfun$run$1(AsyncEventQueue.scala:96)"
78 = {StackTraceElement@23745} "org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1404)"
79 = {StackTraceElement@23746} "org.apache.spark.scheduler.AsyncEventQueue$$anon$2.run(AsyncEventQueue.scala:96)"
```

### PR DESCRIPTION
**Pull request by cerveada**
fixes #101

`Ilegal cyclic reference` can be caused by multiple methods from scala reflection API, this PR moves the whole scala reflection code block into Try block to work around the issue.

### CODE REVIEW

File: src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala [hunk 1]
```diff
@@ -136,6 +136,14 @@ class ReflectionUtilsSpec extends AnyFlatSpec with Matchers with MockitoSugar {
     ReflectionUtils.extractFieldValue[Boolean](Foo, "x") should be(Foo.x)
   }
 
+  // this class must be outside the test to trigger the exception
+  case class WrappingClass(x: ClassWithCyclicAnnotation)
```
Comment (wajda) [inline src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala:140]: Minor: another such dummy classes are located in the test companion object below. I would move this one there as well for the sake of consistency.

File: src/test/scala/za/co/absa/commons/reflect/ClassWithCyclicAnnotation.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.commons.reflect
+
+import za.co.absa.commons.reflection.CyclicAnnotationExample
```
Comment (wajda) [inline src/test/scala/za/co/absa/commons/reflect/ClassWithCyclicAnnotation.scala:19]: I just realized a minor inconsistency we have in a package name - reflection vs reflect. Could you fix it in the scope of this PR please?

```suggestion
import za.co.absa.commons.reflect.CyclicAnnotationExample
```

File: src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala [hunk 3]
```diff
@@ -136,6 +136,14 @@ class ReflectionUtilsSpec extends AnyFlatSpec with Matchers with MockitoSugar {
...
+
+  it should "fallback to Java reflection when Scala one fails via return type" in {
+    val obj = WrappingClass(new ClassWithCyclicAnnotation)
+    ReflectionUtils.extractFieldValue[ClassWithCyclicAnnotation](obj, "x") should not be null
```
Comment (wajda) [inline src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala:144]: just to shut Codacy up...
```suggestion
    val obj = WrappingClass(new ClassWithCyclicAnnotation)
    // scalastyle:off null
    ReflectionUtils.extractFieldValue[ClassWithCyclicAnnotation](obj, "x") should not be null
    // scalastyle:on null
```

File: src/main/scala/za/co/absa/commons/reflect/FieldValueExtractor.scala [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+import za.co.absa.commons.annotation.DeveloperApi
+import za.co.absa.commons.reflect.ReflectionUtils.allInterfacesOf
+
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe.{Mirror, runtimeMirror}
+
+@DeveloperApi
+class FieldValueExtractor[A: ClassTag, B](o: AnyRef, fieldName: String) {
```
Comment (wajda) [inline src/main/scala/za/co/absa/commons/reflect/FieldValueExtractor.scala:27]: Does it need to a class? It doesn't seem to have any state, so shouldn't it rather be an `object` ?

File: src/test/scala/za/co/absa/commons/reflect/ClassWithCyclicAnnotation.scala [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+@CyclicAnnotationExample.CyclicAnnotation
+class ClassWithCyclicAnnotation {
+
+}
```
Comment (wajda) [inline src/test/scala/za/co/absa/commons/reflect/ClassWithCyclicAnnotation.scala:24]: nit: redundant curly braces - Codacy [spotted](https://app.codacy.com/gh/AbsaOSS/commons/file/73868419675/issues/source?bid=15994054&fileBranchId=30255313#l22) it right

```suggestion
class ClassWithCyclicAnnotation
```

File: src/main/scala/za/co/absa/commons/reflect/FieldValueExtractor.scala (hunk 4 above)
Comment (cerveada) [reply src/main/scala/za/co/absa/commons/reflect/FieldValueExtractor.scala:27]: I was trying to reduce the nesting, by defining the o and fieldName as constructor parameters all the methods can be on top level.

**Review by wajda — APPROVED** [review]
LGTM

**Review by dk1844 — APPROVED** [review]
LGTM (just read the code)

**Pull request merged**